bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -1272,24 +1273,46 @@ private boolean needsBuild(InternalBuilder builder, int trigger) {
 		});
 
 		//search for the builder's project
-		if (currentDelta.findNodeAt(builder.getProject().getFullPath()) != null) {
-			if (Policy.DEBUG_BUILD_NEEDED)
-				Policy.debug(toString(builder) + " needs building because of changes in: " + builder.getProject().getName()); //$NON-NLS-1$
-			return true;
+		boolean ret = false;
+		AbstractDataTreeNode found;
+		if ((found = currentDelta.findNodeAt(builder.getProject().getFullPath())) != null) {
+			ret = true;
+			if (!debugEachBuildingBecauseOf(builder, found))
```

Comment:
It is not common to change logic based on calls to debug code. 

Even if the code today does almost same as before, this may lead to changes in debug code that will lead to changes in production code. Additionally, the resulting code as it is now is less clear to understand.

Same below.
It be OK if the debug statements would not interfere with the non debug logic.

Reply:
Ack
Reorganized a little into dedicated debug methods, with minor change to existing behavior (allows for simpler code):
Debug _each_ project with relevant delta, even when not tracing the particular delta nodes.